save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/AgentsController.kt

```diff
@@ -39,10 +40,19 @@ class AgentsController(private val agentStatusRepository: AgentStatusRepository,
      * @param agentStates list of [AgentStatus]es to update in the DB
      */
     @PostMapping("/updateAgentStatusesWithDto")
+    @Transactional
     fun updateAgentStatusesWithDto(@RequestBody agentStates: List<AgentStatusDto>) {
-        agentStates.forEach {
-            val agent = getAgentByContainerId(it.containerId)
-            agentStatusRepository.save(AgentStatus(it.time, it.state, agent))
+        agentStates.forEach { dto ->
+            val agentStatus = agentStatusRepository.findTopByAgentContainerIdOrderByEndTimeDesc(dto.containerId)
+            if (agentStatus != null && agentStatus.state == dto.state) {
+                // updating time
+                agentStatus.endTime = dto.time
+                agentStatusRepository.save(agentStatus)
+            } else {
+                // insert new agent status
+                val agent = getAgentByContainerId(dto.containerId)
+                agentStatusRepository.save(AgentStatus(LocalDateTime.now(), dto.time, dto.state, agent))
```

Comment:
```suggestion
                agentStatusRepository.save(AgentStatus(dto.time, dto.time, dto.state, agent))
```
It makes more sense to use timestamps of status receiving as initially provided by agent

Reply:
Fixed